src/overlap_check/ProgressBar.cpp

```diff
@@ -0,0 +1,78 @@
+
+
+#include "ProgressBar.hpp"
+
+#include <sstream>
+#include <iostream>
+#include <iomanip>
+
+#if defined(__unix__) || defined(__unix) || (defined(__APPLE__) && defined(__MACH__))
+#include <unistd.h>
+#endif
+
+#define BAR_WIDTH 72
+
+ProgressBar::ProgressBar() {
+  // initialize bar
+  set_value(0.0);
+}
+
+ProgressBar::~ProgressBar() {
+  std::cout << "\n";
+}
+
+bool ProgressBar::is_terminal() {
+#ifdef _POSIX_VERSION
+  return isatty(STDOUT_FILENO) != 0;
+#else
+  return false;
+#endif
+}
+
+
+void
+ProgressBar::set_value(double val) {
+
+  std::string bar;
+
+  if (!is_terminal())
+    return;
+
+  // set the bar percentage
+  if (val >= 100.0) {
+    bar.append("100");
+  } else if (val <= 0.0) {
+    bar.append("  0");
+  } else {
+    std::stringstream ss;
+    ss << std::setfill(' ') << std::setw(3) << (int)val;
+    bar.append(ss.str());
+  }
+
+  bar.append("% |");
+  // remaining width of the bar
+  int remaining_width = BAR_WIDTH - bar.size() - 2;
+
+  // set the bar width
+  if (val >= 100.0) {
+    bar.append(remaining_width, '=');
+  } else if (val < 0.0) {
+    bar.append(remaining_width, ' ');
+  } else {
+    int width = (int)((double)remaining_width * val / 100);
+    bar.append(width, '=');
+    bar.append(1, '>');
+    bar.append(remaining_width - width - 1, ' ');
+  }
```

Comment:
Do you need the if/else if /else ? 
could you not at the beginning do something like:
`if val > 100 > val =100`
`else if val < 0 val = 0`

and then just:
```
std::stringstream ss;
ss << std::setfill(' ') << std::setw(3) << (int)val;
bar.append(ss.str());
```
 and then
```
int width = (int)((double)remaining_width * val / 100);
bar.append(width, '=');
bar.append(1, '>');
bar.append(remaining_width - width - 1, ' ');
```

without special cases ?

Reply:
The problem with that is that when we reach 100% `remaining_width` will equal `width` and the final `bar.append` call will have an argument of `-1` (`bar.append(-1, ' ')`). So I could put another if statement there, but I think that what exists delineates the three cases we're concerned with more clearly. Also, this would technically result in move overall calls to `str::append` if we're concerned about performance.

Reply:
Not sure I understand why the -1 is there at all? 

What about:
```
val = max(min(100,val),0);
int width = (int)((double)remaining_width * val / 100);
bar.append(width, '=');
bar.append(remaining_width - width , ' ');
```

Reply:
The `-1` is for the arrowhead (`bar.append(1, '>')`) which I at some point removed accidentally.

I do like the `val = max(min(100.0, val), 0.0)` update quite a bit and moving that `-1` to another location (basically lumping any extra space needed for the prefix, suffix, and arrowhead into the `remaining_width` initialization) let's us get rid of a lot of those `if`s as @bam241 suggested. This comes at the cost of the arrowhead still appearing when the bar is full rather than getting replaced with another `=` character, but the samll aesthetic change doesn't bother me much.

Reply:
Haven't had a chance to check out if there are any concurrecy performance-killers with this update. My sense is that the early exits will avoid those but I'll double-check that tomorrow.

Reply:
No appreciable difference in performance that I could see!